Add Nim-lang support via nim-langserver

a5m0 opened the pull request:
## Summary

- Implement a `NimLanguageServer` class that handles downloading/installing `nimlangserver`, sets up environment variables, writes auxiliary Nim/nimsuggest config files, and improves logging/readiness checks so the process starts reliably.
- Wire the new server into Serena’s language registry and document official Nim support in the README and changelog.
- Provide pytest coverage plus a sample Nim project that exercises symbols, definitions, references, and completions, along with a dedicated nim pytest marker.

## Testing

- `uv run pytest -m nim` (passes locally)

MischaPanch reviewed (changes requested):
Thanks for the contribution! A good step towards adding nim support, but a bit of "polishing" work is still needed. I haven't reviewed all details yet (in particular, the logic of the while loop or the config writing), and will do so once these first comments are addressed.

a5m0 committed:
feat: Add Nim language server support

Rebuild Nim language support on current upstream (origin/main) using
the latest architecture patterns (LanguageServerProcess, self-wiring
Language enum, updated SolidLanguageServer init signature).

Includes a custom NimLanguageServerProcess that sends only
Content-Length headers — nimlangserver 1.12.0 still cannot parse
the Content-Type header that the standard implementation sends.

- NIM enum in Language with source matchers and factory wiring
- NimLanguageServer with nimsuggest config generation, port health
  tracking, and retry-with-backoff for document symbol requests
- Test suite (8 tests): symbols, types, references, goto-def,
  completions across main.nim, utils.nim, types.nim
- nim pytest marker in pyproject.toml

a5m0 committed:
fix: address PR review feedback for Nim language server

- Remove silent-pass guards in tests; use hard assertions
- Fix incorrect line numbers in goto_definition and completion tests
- Fix completion test to match actual SolidLanguageServer return format
- Use shutil.which() for cross-platform nimlangserver discovery
- Also check ~/.nimble/bin when nimble bin isn't on PATH
- Extract config file content to module-level constants
- Expand class and method docstrings explaining design decisions

a5m0 committed:
fix(nim): skip Nim tests when toolchain is not installed

Nim tests were failing with RuntimeError on CI runners (Windows,
Ubuntu, macOS) because the Nim toolchain is not pre-installed.
Add skipif decorator to gracefully skip when neither nim nor
nimlangserver is found in PATH.

a5m0 committed:
fix(test): stabilize flaky Windows CI tests

Increase pyright startup timeout from 5s to 30s in
test_create_with_index_flag, and add Kotlin to the flaky language
server skip list in test_find_symbol, both of which were timing out
on Windows CI runners.

a5m0 committed:
fix(nim): address review feedback with factory method, thread safety, and test improvements

- Extract _create_server_process() factory method in SolidLanguageServer
  so NimLanguageServer can override it cleanly instead of the
  stop-and-replace hack that created then immediately discarded a
  LanguageServerProcess
- Use threading.Event for _has_port_error and _nimsuggest_functional
  flags to ensure thread-safe reads/writes between notification
  handlers and request threads
- Upgrade config file creation logs from info to warning with
  .gitignore suggestion
- Reduce retry delays from [5, 15, 30]s to [2, 5, 10]s
- Replace hardcoded line/column numbers in goto_definition and
  completions tests with dynamic content lookup
- Add trailing newlines to all Nim test repo files
- Add unit tests for _send_payload (Content-Length without
  Content-Type) and _create_nim_config_if_needed (7 edge cases)

MischaPanch commented on src/solidlsp/language_servers/nim_language_server.py:
```diff
@@ -0,0 +1,555 @@
+            log.debug("Could not create config files: %s", e)
+
+    @classmethod
+    def _setup_runtime_dependencies(cls, config: LanguageServerConfig, solidlsp_settings: SolidLSPSettings) -> str:
```
pls use the newer `LanguageServerDependencyProvider` interface for downloading dependencies, all new LS should build on that. You can find many examples for how to use it in the exiting LS (it's usually a very small refactoring to change existing code to use the new interface)

MischaPanch commented on test/solidlsp/nim/test_nim_basic.py:
```diff
@@ -0,0 +1,199 @@
+        ref_paths = [ref.get("relativePath", "") for ref in refs]
+        assert (
+            any("utils.nim" in p for p in ref_paths) or refs == []
+        ), f"If references are returned, utils.nim (definition site) should be among them, got: {ref_paths}"
```
this should assert that found a reference in a different file, but it only checks the same file as the source symbol

a5m0 replied:
test_find_references_across_files (lines 150-181) now:
  - Finds formatNumber call in main.nim (defined in utils.nim)
  - Asserts request_references returns results including main.nim
  - Also calls request_definition and asserts it resolves to utils.nim (cross-file verification)

Additionally, a dedicated test_request_referencing_symbols_cross_file test (lines 249-275) queries formatNumber's definition in utils.nim and asserts that request_referencing_symbols finds a reference in main.nim.

MischaPanch commented on test/solidlsp/nim/test_nim_unit.py:
```diff
@@ -0,0 +1,155 @@
+"""
+
+import json
+import os
```
I think this module should be deleted. I don't see the value in these tests, and they use private methods (which no test should do)

a5m0 committed:
fix(nim): address review feedback and fix nimsuggest project detection

- Extract shared _retry_on_empty helper for hover/references/definition
  retries, replacing three near-identical retry loops
- Add _detect_nim_project_mapping to inject projectMapping into LSP init
  options, ensuring nimsuggest starts with the correct .nim entry point
  instead of config.nims (fixes flaky cross-file references/goto-def)
- Add request cap (200) and debug logging to goto-definition fallback
- Filter goto-definition fallback to .nim files only
- Add test_repo.nimble and nim.cfg to test repo for proper project detection
- Fix broken ../src path in test repo config.nims
- Add hover test, delete test_nim_unit.py (tested private methods)

MischaPanch commented on src/solidlsp/language_servers/nim_language_server.py:
```diff
@@ -0,0 +1,1056 @@
+        return last_content_line
+
+    @override
+    def request_document_symbols(self, relative_file_path: str, file_buffer: "LSPFileBuffer | None" = None) -> "DocumentSymbols":
```
You override both the public and the private method, is that really needed?

a5m0 commented on src/solidlsp/language_servers/nim_language_server.py:
```diff
@@ -0,0 +1,1056 @@
+
+        overview = self.request_overview("")
+
+        cap_reached = False
```
Unfortunately the tooling available is not good (part of what encouraged me to add Nim to Serena), and I agree that the current nim-langserver support should be marked as experimental. I will try to address your new comments, but ultimately I expect Nim support to be sub-par until https://github.com/nim-lang/nimony is ready.

zerone0x reviewed (approved):
Pulled the branch locally to make sure nothing obvious regressed. The new `NimLanguageServerProcess` and config helpers are wired through `Language.NIM`, the factory override looks clean, and the docs/changelog entries are in place. Ran `uv run pytest -m nim` — everything skips on this machine because neither Nim nor nimlangserver is installed, but there were no unexpected failures. From my side this looks good; thanks for pushing the full compatibility layer over the finish line.

zerone0x commented:
I compared this PR with #1150 as requested.

Summary: #658 is the right base to continue with. #1150 is much smaller/simpler, but it does not cover the nimlangserver-specific problems that this PR already handles.

What #658 has that #1150 lacks:

- Dependency handling via `LanguageServerDependencyProviderSinglePath`, including discovery in `~/.nimble/bin`, PATH setup for `nimsuggest`, and optional `nimble install nimlangserver`.
- A custom `NimLanguageServerProcess` that sends only `Content-Length`. #1150 uses the default process, so it likely still hits the nimlangserver `Content-Type` parsing issue discussed here.
- Project/config handling: `nimsuggest.cfg`, `nim.cfg`, and `projectMapping` detection to keep nimsuggest pointed at a real `.nim` entry point instead of e.g. `config.nims`.
- Readiness/status handling for `extension/statusUpdate`, `port=0`, and nimsuggest restart/warmup states.
- Workarounds for the currently weak LS behavior: symbol range extension, retries for symbols/hover/definition/references, and the guarded goto-definition fallback for cross-file referencing symbols.
- Broader tests: symbols, types, goto-definition, references, completions, hover, and `request_referencing_symbols`. #1150 mainly tests basic symbols plus two direct cross-file reference checks.

What might be reusable from #1150:

- The simpler test repo shape (`src/calculator.nim`, `src/utils.nim`, `main.nim`) is a bit more conventional, but #658's current fixture is intentionally richer and better exercises the hacks/fallbacks.
- The user-facing install wording in #1150 is shorter, but #658's dependency provider makes manual installation less central.

I do not see any core implementation from #1150 that should replace parts of this PR. The remaining work looks more like polishing #658 than merging/cherry-picking #1150.

Remaining items I would prioritize before merge, based on the latest review thread:

1. Mark Nim support as experimental in user-facing docs (README language list and `docs/01-about/020_programming-languages.md`), similar to Scala/Kotlin expectation-setting.
2. Add a short Nim note documenting the current nimlangserver/nimsuggest limitations: generated `nim.cfg`/`nimsuggest.cfg`, `Content-Type` workaround, range-fix heuristic, retries/fallbacks, and possible incomplete cross-file references.
3. Optionally mention in `CHANGELOG.md` that initial Nim support is experimental.
4. If touching code again, address the small review nits that are straightforward; otherwise I think the important behavior differences vs #1150 are already covered here.

I also ran `uv run --extra dev pytest -m nim` locally on this branch; the Nim tests were all skipped because this machine does not have the Nim toolchain/nimlangserver installed, but collection succeeded and there were no unexpected failures.

MischaPanch commented:
Seems like the nim community is not particularly eager in this regard. Closing as stale
